### ISSUE TITLE
Should SIM105 `UseContextlibSuppress` exempt try blocks with unreachable end?

### ISSUE THREAD

**Issue by andersk** [charliermarsh/ruff#1947]
This code violates SIM105 (Use `contextlib.suppress(OSError)` instead of try-except-pass):

```python
for i in range(1, 100):
    try:
        realm.string_id = prefix + str(i)
        realm.save(update_fields=["string_id"])
        break
    except IntegrityError:
        pass
else:
    raise RuntimeError(f"Unable to find a good string_id for realm {realm}")
```

But fixing it would make the code more confusing:

```python
for i in range(1, 100):
    with contextlib.suppress(IntegrityError):
        realm.string_id = prefix + str(i)
        realm.save(update_fields=["string_id"])
        break
else:
    raise RuntimeError(f"Unable to find a good string_id for realm {realm}")
```

because this visually suggests that the `for` loop “always” `break`s and it’s not obvious why the `else` can be reached.

If a `try:` block has an unreachable end because all its non-exceptional control flow paths end with a `return`, `raise`, `break`, or `continue`, should we exempt it from SIM105?

**Comment by charliermarsh** [comment]
Yeah I agree that the fixed version is more confusing. (I kind of feel like this rule always makes code more confusing? Though I've been trying to avoid opining too much on rule usefulness while implementing existing, third-party plugins.) I think this suggestion makes sense.

**Comment by andersk** [comment]
Upstream doesn’t have this restriction, but it has a different one: only `try` statements with a [single body statement](https://github.com/MartinThoma/flake8-simplify/blob/0.19.3/flake8_simplify/rules/ast_try.py#L46) are considered. I think this also makes sense, because it may not be immediately apparent that a “suppressed” exception in the first body statement still short-circuits the second body statement.

The upstream restriction doesn’t consider whether the single body statement is itself a compound statement, though, and probably should.

**Comment by charliermarsh** [comment]
Oh, yeah, that looks like an oversight.
